drivers/serial/serial.c

```diff
@@ -1761,7 +1771,7 @@ void uart_connected(FAR uart_dev_t *dev, bool connected)
 void uart_reset_sem(FAR uart_dev_t *dev)
 {
   nxsem_reset(&dev->xmitsem,  0);
-  nxsem_reset(&dev->recvsem,  0);
+  nxsem_reset(&dev->recvsem,  1);
```

Comment:
should we change 0 to dev->recvwaiting?

Reply:
No. It doesn't hurt. It is checked in the receive interrupt. After closing the driver no interrupt will happen.

Reply:
But, if nobody is blocking in the uart_read, dev->recvsem reset to 1 after close, then the next open and then read will wake up immediately with the empty buffer.

Reply:
The close function has already a logic to wait for clearing transmit buffer. So the close function will block.

/* Prevent blocking if the device is opened with O_NONBLOCK */
  if ((filep->f_oflags & O_NONBLOCK) == 0)
    {
      /* Now we wait for the transmit buffer(s) to clear */
      uart_tcdrain(dev, 4 * TICK_PER_SEC);
    }

Reply:
Yes, it will wake up immediately with an empty buffer. But it will check the condition
if (rxbuf->head == rxbuf->tail)
and take the semaphore again.

Reply:
Is it better to avoid the fake wakeup in the first place?

Reply:
Yes, but it will complicate logic. We can do double reset 
nxsem_reset(&dev->recvsem,  1);
nxsem_reset(&dev->recvsem,  0);
But it looks strange.

IMHO One fake wakeup is better than one line of source code.

Reply:
It is more about nxsem_reset function. It knows that sem->semcount is negative, but posts semaphore only if argument is positive. I think something is missing there.

Reply:
The strange is that we always initialize dev->recvsem with 0, but reset it to 1.
What do you think this?
nxsem_reset(&dev->recvsem, dev->recvwaiting);
which reset dev->recvsem to 0 in most case just like init but to 1 in the abnormal condition.

Reply:
OK.

Is boolean true always 1?

Reply:
should be or change to dev->recvwaiting != 0

Reply:
It is also boolean. So it grows +11 lines. :)

Reply:
I am wondering if this is the correct approach. 

Is the the close a cancellation point? Shouldn't that release the waiter? Is the problem somewhere else?

Reply:
What else cancellation point can be? Ioctl?

The close shuts down the low level driver. There will be no any new event from it. As for now, the read function dead locks without new events.

Another approach is to ask low level driver to generate the receive event on shut down. But we have many drivers which don't do it.